File: docs/py/brownianmotion.py
"""2D hard-disc gas with a heavy tracer -> Brownian motion.

Bath particles and the tracer collide elastically as hard discs.
Momentum transfer from the bath kicks the tracer along a random walk.

Physics from "brownianmotion.py"; the animation loop and drawing are done by
the JS layer, which calls step() through web_frame(). The state setup is
wrapped in init() so the web page's Reset button can rerun it.
"""

import numpy as np

kB = 1.380649e-23

# ---------------- parameters ----------------
L = 1.0e-6  # box side, m
T = 1000  # K

n_bath = 180
m_bath = 4.65e-26  # N2 molecule, kg
r_bath = 8.0e-9  # inflated ~50x over real N2 so collisions are frequent

M_trac = 10 * m_bath
R_trac = 8.0e-8

dt = 1.0e-12  # s
spf = 10  # physics steps per animation frame
seed = 2  # fixed by default so a run is reproducible; the web UI can re-roll it


def init():
    global rng, n, mass, radius, pos, vel

    rng = np.random.default_rng(seed)

    # ---------------- state ----------------
    n = n_bath + 1  #total number of particles: index 0 is the tracer
    mass = np.concatenate(([M_trac], np.full(n_bath, m_bath)))
    radius = np.concatenate(([R_trac], np.full(n_bath, r_bath)))

    # grow the lattice until enough sites survive the tracer exclusion, otherwise
    # a big tracer or a large n_bath silently leaves pos shorter than mass/radius
    k = int(np.ceil(np.sqrt(n_bath * 1.6)))
    while True:
        gx, gy = np.meshgrid(np.linspace(0.05 * L, 0.95 * L, k), np.linspace(0.05 * L, 0.95 * L, k))
        spawnsites = np.column_stack([gx.ravel(), gy.ravel()]) #every possible site of current grid
        spawnsites = spawnsites[np.hypot(*(spawnsites - L / 2).T) > R_trac + 2 * r_bath] #sites clear of tracer
        if len(spawnsites) >= n_bath:
            break
        k += 2
    spawnsites = rng.permutation(spawnsites)[:n_bath] #chose some random bath spawn sites

    pos = np.vstack([[L / 2, L / 2], spawnsites]) #[L/2,L/2] is default tracer origin, so we just stack that with spawnsites to get a full array of all the initial particle positions

    # velocities: each species from its own Maxwell-Boltzmann
    vel = np.empty((n, 2)) #n rows for n particles, each row contains vx and vy
    vel[0] = rng.normal(0.0, np.sqrt(kB * T / M_trac), 2)  #tracer velocity    #syntax: rng.normal(mean, std, size of output)
    vel[1:] = rng.normal(0.0, np.sqrt(kB * T / m_bath), (n_bath, 2)) #1: - 1 onwards. bath particle velocities.
    vel -= (mass[:, None] * vel).sum(0) / mass.sum()  # calculate centre of mass velocity and subtract from all velocities, such that we are in the zero momentum frame at initialisation, so gas particles don't all drift apart immediately.


# ---------------- physics ----------------
def step():
    global pos  # needed: `pos +=` is an assignment, so Python would make it local
    pos += vel * dt #moves every particle in a straight line at constant velocity per step

    # walls: reflect, using abs() so a disc still overlapping can't double-flip
    lo = pos - radius[:, None] < 0.0
    hi = pos + radius[:, None] > L
    vel[lo] = np.abs(vel[lo])
    vel[hi] = -np.abs(vel[hi])
    np.clip(pos, radius[:, None], L - radius[:, None], out=pos)

    # pair overlaps: vectorised O(n^2) detection
    d = pos[:, None, :] - pos[None, :, :]
    dist2 = np.einsum('ijk,ijk->ij', d, d)
    Rsum = radius[:, None] + radius[None, :]
    ia, ib = np.nonzero(np.triu(dist2 < Rsum ** 2, 1))

    for a, b in zip(ia, ib):
        dist = np.sqrt(dist2[a, b])
        nhat = d[a, b] / dist  # unit vector from b to a
        vn = (vel[a] - vel[b]) @ nhat
        if vn >= 0.0:
            continue  # already separating, leave alone

        mu = mass[a] * mass[b] / (mass[a] + mass[b])
        J = 2.0 * mu * vn
        vel[a] -= (J / mass[a]) * nhat
        vel[b] += (J / mass[b]) * nhat

        # separate them, displacement inversely proportional to mass
        overlap = Rsum[a, b] - dist
        fa = mass[b] / (mass[a] + mass[b])
        pos[a] += fa * overlap * nhat
        pos[b] -= (1.0 - fa) * overlap * nhat


def kinetic_energy():
    return 0.5 * (mass * (vel ** 2).sum(1)).sum()


def momentum():
    return (mass[:, None] * vel).sum(0)


init()


# ---- adapter for the web UI ----

def web_setup(n_bath_=n_bath, m_bath_=m_bath, r_bath_=r_bath,
              M_trac_=M_trac, R_trac_=R_trac, T_=T, seed_=seed):
    # defaults are bound at def time to the module values above, so a bare
    # web_setup() still rebuilds the standard run
    global n_bath, m_bath, r_bath, M_trac, R_trac, T, seed
    n_bath, seed = int(n_bath_), int(seed_)   # range inputs arrive as JSON floats
    m_bath, r_bath = float(m_bath_), float(r_bath_)
    M_trac, R_trac = float(M_trac_), float(R_trac_)
    T = float(T_)
    init()
    return {
        "L": L,
        "T": T,
        "dt": dt,          # so the UI can turn frames into simulated time
        "n_bath": n_bath,
        "m_bath": m_bath,
        "M_trac": M_trac,
        "R_trac": R_trac,
        "r_bath": r_bath,
        "tracer": pos[0].tolist(),
        "bath": pos[1:].tolist(),
    }


def web_frame(substeps):
    for _ in range(int(substeps)):
        step()
    return {
        "tracer": pos[0].tolist(),
        "bath": np.round(pos[1:], 10).tolist(),   # 0.1 nm precision, well under a pixel
    }

File: docs/js/brownianmotion.js
// Brownian motion — UI for py/brownianmotion.py.
// The hard-disc physics runs in Python; this file only animates the results.

PROJECTS.push({
  id: 'brownian',
  name: '2. Brownian Motion',
  blurb: 'A 2D animation of a large tracer particle, initialised at the centre of the grid, with smaller "bath" particles initialised at randomised positions on a square grid, clear of the tracer. Collisions with walls reverse the component perpendicular to the wall only. Collisions between particles are elastic; energy is conserved to double floating point precision over 5000 steps. Initial bath particle velocity components are sampled from Boltzmann distributions depending on the temperature and particle masses that you can set manually. Time step = 1 picosecond, box width = 1 micrometre. There are 10 timesteps per animation frame. The radii of the particles are inflated compared to actual gas particles in the context of the box size, so that the collisions can actually be seen. You can click save to save the last walk, and automatically start a new one. ',

  async render(page) {
    const SIZE = 520;
    const STEPS_PER_FRAME = 10;   // matches spf in the python module
    const U = 1.66053907e-27;     // atomic mass unit, kg

    // saved walks, drawn under the live one. blue is the bath colour, so it is out
    const SAVED_COLOURS = ['#2ca02c', '#9467bd', '#ff7f0e', '#8c564b', '#e377c2'];

    // slider state, in display units; params() converts to SI for python
    let nBath = 180, mBathU = 28, rBathNm = 8, tracerExp = 1, rTracNm = 80, temp = 1000;
    let seed = 2;

    const params = () => {
      const mBath = mBathU * U;
      return [nBath, mBath, rBathNm * 1e-9, (10 ** tracerExp) * mBath, rTracNm * 1e-9,
              temp, seed];
    };
    const newSeed = () => Math.floor(Math.random() * 1e6);

    const canvas = el('canvas', { width: SIZE, height: SIZE,
      style: { background: '#fff', border: '1px solid #e4e4e7' } });
    const ctx = canvas.getContext('2d');

    // saved walks never change once saved, so they are stroked once into an
    // offscreen canvas and blitted per frame rather than re-stroked every frame
    const savedCanvas = el('canvas', { width: SIZE, height: SIZE });
    const savedCtx = savedCanvas.getContext('2d');
    let savedCount = 0;

    let setup = await pyCall('brownianmotion.web_setup', ...params());
    const SCALE = SIZE / setup.L;
    let bath = setup.bath;
    let trail = [setup.tracer];

    function strokeTrail(c, pts, colour) {
      if (pts.length < 2) return;
      c.strokeStyle = colour;
      c.lineWidth = 1;
      c.beginPath();
      c.moveTo(pts[0][0] * SCALE, SIZE - pts[0][1] * SCALE);
      for (let i = 1; i < pts.length; i++) {
        c.lineTo(pts[i][0] * SCALE, SIZE - pts[i][1] * SCALE);
      }
      c.stroke();
    }

    function draw() {
      ctx.fillStyle = '#fff';
      ctx.fillRect(0, 0, SIZE, SIZE);

      ctx.drawImage(savedCanvas, 0, 0);
      strokeTrail(ctx, trail, 'rgba(214,39,40,0.6)');

      ctx.fillStyle = 'rgba(31,119,180,0.7)';
      for (const p of bath) {
        ctx.beginPath();
        ctx.arc(p[0] * SCALE, SIZE - p[1] * SCALE, setup.r_bath * SCALE, 0, 2 * Math.PI);
        ctx.fill();
      }

      const t = trail[trail.length - 1];
      ctx.fillStyle = '#d62728';
      ctx.beginPath();
      ctx.arc(t[0] * SCALE, SIZE - t[1] * SCALE, setup.R_trac * SCALE, 0, 2 * Math.PI);
      ctx.fill();
    }

    let running = true;
    let alive = true;
    let gen = 0;       // a re-init invalidates any frame still in flight
    let simTime = 0;   // simulated time, seconds — not wall clock

    // tabular-nums keeps the digits from jittering as the clock runs
    const clock = el('span', { class: 'val' });
    const showClock = () => { clock.textContent = 't = ' + (simTime * 1e9).toFixed(2) + ' ns'; };

    async function loop() {
      while (alive) {
        if (running) {
          const my = gen;
          const frame = await pyCall('brownianmotion.web_frame', STEPS_PER_FRAME);
          if (my === gen) {
            bath = frame.bath;
            trail.push(frame.tracer);
            if (trail.length > 6000) trail.shift();
            simTime += STEPS_PER_FRAME * setup.dt;
            showClock();
            draw();
          }
        }
        await new Promise(r => requestAnimationFrame(r));
      }
    }

    const note = el('span', { class: 'note' });

    function noteText() {
      return `${setup.n_bath} bath particles at ${(setup.m_bath / U).toFixed(0)} u, ` +
             `tracer mass ${(setup.M_trac / setup.m_bath).toPrecision(2)}× bath, ` +
             `T = ${setup.T} K, box ${setup.L * 1e6} μm (radii inflated)`;
    }

    async function reinit() {
      const my = ++gen;
      try {
        const s = await pyCall('brownianmotion.web_setup', ...params());
        if (my !== gen) return;
        setup = s;
        bath = s.bath;
        trail = [s.tracer];
        simTime = 0;      // fresh run, fresh clock
        showClock();
        note.textContent = noteText();
        draw();
      } catch (err) {
        // a throw in a control handler would otherwise vanish into the console
        note.textContent = 'setup failed: ' + err.message;
        console.error(err);
      }
    }

    const playBtn = el('button', { class: 'btn', onclick: () => {
      running = !running;
      playBtn.textContent = running ? 'Pause' : 'Play';
    } }, 'Pause');

    // saving without re-rolling the seed would replay the same walk straight on top
    const saveBtn = el('button', { class: 'btn', onclick: () => {
      savedCtx.globalAlpha = 0.45;
      strokeTrail(savedCtx, trail, SAVED_COLOURS[savedCount % SAVED_COLOURS.length]);
      savedCtx.globalAlpha = 1;
      savedCount++;
      seed = newSeed();
      reinit();
    } }, 'Save walk');

    const randBtn = el('button', { class: 'btn', onclick: () => {
      seed = newSeed();
      reinit();
    } }, 'Randomise bath');

    const resetBtn = el('button', { class: 'btn', onclick: () => {
      savedCtx.clearRect(0, 0, SIZE, SIZE);
      savedCount = 0;
      seed = 2;         // back to the reproducible default run
      reinit();
    } }, 'Reset');

    const sliders = [
      slider({ label: 'Bath particles', min: 20, max: 400, step: 10, value: nBath,
        fmt: v => String(v), oninput: v => { nBath = v; } }),
      slider({ label: 'Bath mass', min: 8, max: 120, step: 1, value: mBathU,
        fmt: v => v + ' u', oninput: v => { mBathU = v; } }),
      slider({ label: 'Bath radius', min: 6, max: 16, step: 0.5, value: rBathNm,
        fmt: v => v + ' nm', oninput: v => { rBathNm = v; } }),
      slider({ label: 'Tracer mass', min: -2, max: 2, step: 0.05, value: tracerExp,
        fmt: v => (10 ** v).toPrecision(2) + '× bath', oninput: v => { tracerExp = v; } }),
      slider({ label: 'Tracer radius', min: 20, max: 200, step: 5, value: rTracNm,
        fmt: v => v + ' nm', oninput: v => { rTracNm = v; } }),
      // capped at 3000 K: speeds go as sqrt(T) against a fixed dt, and above that
      // the fixed step starts to miss collisions at the small-radius end
      slider({ label: 'Temperature', min: 100, max: 3000, step: 50, value: temp,
        fmt: v => v + ' K', oninput: v => { temp = v; } })
    ];
    // readouts follow the drag, but rebuilding the whole gas is far too heavy
    // to do per pixel, so the re-init waits for release
    for (const s of sliders) s.inp.addEventListener('change', () => reinit());

    page.append(el('div', { class: 'controls' },
                  playBtn, saveBtn, randBtn, resetBtn,
                  ...sliders.map(s => s.root), note),
                // .canvas-row is the existing flex helper used by diffraction and
                // orbitals; it puts the clock beside the box without new CSS
                el('div', { class: 'card' },
                  el('div', { class: 'canvas-row' }, canvas, clock)));

    note.textContent = noteText();
    showClock();
    draw();
    loop();

    return () => { alive = false; };   // stop the loop when leaving the page
  }
});
